pkg/oauthserver/oauth/external/gitlab/gitlab.go

```diff
@@ -133,3 +119,49 @@ func (p *provider) GetUserIdentity(data *osincli.AccessData) (authapi.UserIdenti
 
 	return identity, true, nil
 }
+
+func (p *provider) getUserData(token string) (*gitlabUser, error) {
+	// try the v4 API first
+	userdata, err := p.getUserDataURL(token, p.userAPIURLV4)
+	if err == nil {
+		return userdata, nil
+	}
+	// fallback to the v3 API
+	return p.getUserDataURL(token, p.userAPIURLV3)
+}
+
+func (p *provider) getUserDataURL(token, url string) (*gitlabUser, error) {
```

Comment:
Sounds like wrapping getUserIdentity with a retry with fallback to V3 only on specific errors and only once and for all would makes for less code changes too.

Reply:
> specific errors

Unclear to me what those errors are.

> only once

Unclear to me how to do that without locking.

Reply:
specific errors -> 404 or other if gitlab returnssomething special for missing endpoints
only once -> do we have a way to atomically replace a pointer ?